File: py-src/minicode/mock_model.py
from __future__ import annotations

import time

from minicode.types import AgentStep


def _last_user_message(messages):
    return next((message["content"] for message in reversed(messages) if message["role"] == "user"), "")


def _last_tool_message(messages):
    return next((message for message in reversed(messages) if message["role"] == "tool_result"), None)


def _latest_assistant_call(messages):
    call = next((message for message in reversed(messages) if message["role"] == "assistant_tool_call"), None)
    return call["toolName"] if call else None


class MockModelAdapter:
    def next(self, messages, on_stream_chunk=None):
        tool_message = _last_tool_message(messages)
        if tool_message and tool_message["role"] == "tool_result":
            last_call = _latest_assistant_call(messages)
            if last_call == "list_files":
                return AgentStep(type="assistant", content=f"Directory contents:\n\n{tool_message['content']}")
            if last_call == "read_file":
                return AgentStep(type="assistant", content=f"File contents:\n\n{tool_message['content']}")
            if last_call in {"write_file", "edit_file", "modify_file", "patch_file"}:
                return AgentStep(type="assistant", content=tool_message["content"])
            return AgentStep(type="assistant", content=f"I received the tool result:\n\n{tool_message['content']}")

        user_text = _last_user_message(messages).strip()
        tool_id = f"mock-{int(time.time() * 1000)}"

        if user_text == "/tools":
            return AgentStep(
                type="assistant",
                content="Available tools: ask_user, list_files, grep_files, read_file, write_file, edit_file, patch_file, run_command",
            )

        if user_text.startswith("/ls"):
            directory = user_text.replace("/ls", "", 1).strip()
            return AgentStep(
                type="tool_calls",
                calls=[{"id": tool_id, "toolName": "list_files", "input": {"path": directory} if directory else {}}],
            )

        if user_text.startswith("/grep "):
            payload = user_text[len("/grep ") :].strip()
            pattern, _, search_path = payload.partition("::")
            return AgentStep(
                type="tool_calls",
                calls=[
                    {
                        "id": tool_id,
                        "toolName": "grep_files",
                        "input": {"pattern": pattern.strip(), "path": search_path.strip() or None},
                    }
                ],
            )

        if user_text.startswith("/read "):
            return AgentStep(
                type="tool_calls",
                calls=[{"id": tool_id, "toolName": "read_file", "input": {"path": user_text[len('/read ') :].strip()}}],
            )

        if user_text.startswith("/cmd "):
            payload = user_text[len("/cmd ") :].strip()
            return AgentStep(type="tool_calls", calls=[{"id": tool_id, "toolName": "run_command", "input": {"command": payload}}])

        if user_text.startswith("/write "):
            payload = user_text[len("/write ") :]
            target_path, separator, content = payload.partition("::")
            if not separator:
                return AgentStep(type="assistant", content="Usage: /write <path>::<content>")
            return AgentStep(
                type="tool_calls",
                calls=[{"id": tool_id, "toolName": "write_file", "input": {"path": target_path.strip(), "content": content}}],
            )

        if user_text.startswith("/edit "):
            payload = user_text[len("/edit ") :]
            parts = payload.split("::")
            if len(parts) != 3:
                return AgentStep(type="assistant", content="Usage: /edit <path>::<search>::<replace>")
            target_path, search, replace = parts
            return AgentStep(
                type="tool_calls",
                calls=[{"id": tool_id, "toolName": "edit_file", "input": {"path": target_path.strip(), "search": search, "replace": replace}}],
            )

        if user_text.startswith("/patch "):
            payload = user_text[len("/patch ") :]
            parts = payload.split("::")
            if len(parts) < 3 or len(parts) % 2 == 0:
                return AgentStep(type="assistant", content="Usage: /patch <path>::<search1>::<replace1>::<search2>::<replace2> ...")
            target_path, *ops = parts
            replacements = []
            for index in range(0, len(ops), 2):
                replacements.append({"search": ops[index], "replace": ops[index + 1]})
            return AgentStep(
                type="tool_calls",
                calls=[{"id": tool_id, "toolName": "patch_file", "input": {"path": target_path.strip(), "replacements": replacements}}],
            )

        return AgentStep(
            type="assistant",
            content="\n".join(
                [
                    "This is a minimal MiniCode Python shell.",
                    "You can try:",
                    "/tools",
                    "/ls",
                    "/grep pattern::src",
                    "/read README.md",
                    "/cmd pwd",
                    "/write notes.txt::hello",
                    "/edit notes.txt::hello::hello world",
                ]
            ),
        )

File: ts-src/src/mock-model.ts
import type { AgentStep, ChatMessage, ModelAdapter } from './types.js'

function lastUserMessage(messages: ChatMessage[]): string {
  const last = [...messages].reverse().find(message => message.role === 'user')
  return last?.content ?? ''
}

function lastToolMessage(messages: ChatMessage[]): ChatMessage | undefined {
  return [...messages].reverse().find(message => message.role === 'tool_result')
}

function extractLatestAssistantCall(messages: ChatMessage[]): string | undefined {
  const last = [...messages]
    .reverse()
    .find(
      message =>
        message.role === 'assistant_tool_call',
    )
  return last?.role === 'assistant_tool_call'
    ? last.toolName
    : undefined
}

export class MockModelAdapter implements ModelAdapter {
  async next(messages: ChatMessage[]): Promise<AgentStep> {
    const toolMessage = lastToolMessage(messages)
    if (toolMessage?.role === 'tool_result') {
      const lastCall = extractLatestAssistantCall(messages)
      if (lastCall === 'list_files') {
        return {
          type: 'assistant',
          content: `目录内容如下：\n\n${toolMessage.content}`,
        }
      }

      if (lastCall === 'read_file') {
        return {
          type: 'assistant',
          content: `文件内容如下：\n\n${toolMessage.content}`,
        }
      }

      if (lastCall === 'write_file' || lastCall === 'edit_file') {
        return {
          type: 'assistant',
          content: toolMessage.content,
        }
      }

      return {
        type: 'assistant',
        content: `我拿到了工具结果：\n\n${toolMessage.content}`,
      }
    }

    const userText = lastUserMessage(messages).trim()

    if (userText === '/tools') {
      return {
        type: 'assistant',
        content: '可用工具：ask_user, list_files, grep_files, read_file, write_file, edit_file, run_command',
      }
    }

    if (userText.startsWith('/ls')) {
      const dir = userText.replace('/ls', '').trim()
      return {
        type: 'tool_calls',
        calls: [{
          id: `mock-${Date.now()}`,
          toolName: 'list_files',
          input: dir ? { path: dir } : {},
        }],
      }
    }

    if (userText.startsWith('/grep ')) {
      const payload = userText.slice('/grep '.length).trim()
      const [pattern, searchPath] = payload.split('::')
      return {
        type: 'tool_calls',
        calls: [{
          id: `mock-${Date.now()}`,
          toolName: 'grep_files',
          input: {
            pattern: pattern.trim(),
            path: searchPath?.trim() || undefined,
          },
        }],
      }
    }

    if (userText.startsWith('/read ')) {
      return {
        type: 'tool_calls',
        calls: [{
          id: `mock-${Date.now()}`,
          toolName: 'read_file',
          input: { path: userText.slice('/read '.length).trim() },
        }],
      }
    }

    if (userText.startsWith('/cmd ')) {
      const parts = userText.slice('/cmd '.length).trim().split(/\s+/)
      const [command, ...args] = parts
      return {
        type: 'tool_calls',
        calls: [{
          id: `mock-${Date.now()}`,
          toolName: 'run_command',
          input: { command, args },
        }],
      }
    }

    if (userText.startsWith('/write ')) {
      const payload = userText.slice('/write '.length)
      const splitAt = payload.indexOf('::')
      if (splitAt === -1) {
        return {
          type: 'assistant',
          content: '用法: /write 路径::内容',
        }
      }

      return {
        type: 'tool_calls',
        calls: [{
          id: `mock-${Date.now()}`,
          toolName: 'write_file',
          input: {
            path: payload.slice(0, splitAt).trim(),
            content: payload.slice(splitAt + 2),
          },
        }],
      }
    }

    if (userText.startsWith('/edit ')) {
      const payload = userText.slice('/edit '.length)
      const [targetPath, search, replace] = payload.split('::')
      if (!targetPath || search === undefined || replace === undefined) {
        return {
          type: 'assistant',
          content: '用法: /edit 路径::查找文本::替换文本',
        }
      }

      return {
        type: 'tool_calls',
        calls: [{
          id: `mock-${Date.now()}`,
          toolName: 'edit_file',
          input: {
            path: targetPath.trim(),
            search,
            replace,
          },
        }],
      }
    }

    return {
      type: 'assistant',
      content: [
        '这是一个最小骨架版本。',
        '你可以试试：',
        '/tools',
        '/ls',
        '/grep pattern::src',
        '/read README.md',
        '/cmd pwd',
        '/write notes.txt::hello',
        '/edit notes.txt::hello::hello world',
      ].join('\n'),
    }
  }
}
